packages/hebao_v1/test/helpers/Signature.ts

```diff
@@ -0,0 +1,146 @@
+) {
+  let signature: string;
+  switch (+type) {
+    case SignatureType.ETH_SIGN: {
+      const privateKey = getPrivateKey(signer);
+      signature = appendType(await signEthereum(message, privateKey), type);
+      break;
+    }
+    case SignatureType.EIP_712: {
+      const privateKey = getPrivateKey(signer);
+      signature = appendType(await signEIP712(message, privateKey), type);
+      break;
+    }
+    case SignatureType.WALLET: {
+      try {
+        const wallet = await ctx.contracts.BaseWallet.at(signer);
+        const walletOwner = await wallet.owner();
+        const privateKey = getPrivateKey(walletOwner);
+
+        // Sign using the wallet owner
+        signature = appendType(
+          appendType(
+            await signEIP712(message, privateKey),
+            SignatureType.EIP_712
+          ),
+          type
+        );
+      } catch {
+        signature = appendType("", type);
+      }
+      break;
+    }
+    default: {
+      // console.log("Unsupported signature type: " + type);
+      signature = appendType("", type);
+    }
+  }
+  return signature;
+}
+
+export function appendType(str: string, type: SignatureType) {
+  const data = new Bitstream(str);
+  data.addNumber(type, 1);
+  return data.getData();
+}
+
+async function signEthereum(message: Buffer, privateKey: string) {
+  const parts = [
+    Buffer.from("\x19Ethereum Signed Message:\n32", "utf8"),
+    message
+  ];
+  const totalHash = ethUtil.keccak(Buffer.concat(parts));
+  const signature = ethUtil.ecsign(totalHash, new Buffer(privateKey, "hex"));
+
+  const data = new Bitstream();
+  data.addHex(ethUtil.bufferToHex(signature.r));
+  data.addHex(ethUtil.bufferToHex(signature.s));
+  data.addHex(ethUtil.bufferToHex(signature.v));
+  return data.getData();
+}
+
+async function signEIP712(message: Buffer, privateKey: string) {
+  const signature = ethUtil.ecsign(message, new Buffer(privateKey, "hex"));
+
+  const data = new Bitstream();
+  data.addHex(ethUtil.bufferToHex(signature.r));
+  data.addHex(ethUtil.bufferToHex(signature.s));
+  data.addHex(ethUtil.bufferToHex(signature.v));
+  return data.getData();
+}
+
+export async function verifySignatures(
+  signers: string[],
+  message: Buffer,
+  signatures: string[]
+) {
+  assert(signers.length == signatures.length, "invalid input");
+  for (let i = 0; i < signatures.length; i++) {
+    verifySignature(signers[i], message, signatures[i]);
+  }
+}
+
+function verifySignature(signer: string, message: Buffer, signature: string) {
+  const data = new Bitstream(signature);
+  const type = data.extractUint8(data.length() - 1);
+  if (type === SignatureType.ETH_SIGN) {
+    const personalMessage = ethUtil.hashPersonalMessage(message);
+    assert(
+      verifyECDSA(personalMessage, data, signer),
+      "invalid ETH_SIGN signature"
+    );
+  } else if (type === SignatureType.EIP_712) {
+    assert(verifyECDSA(message, data, signer), "invalid EIP_712 signature");
+  }
```

Comment:
Added the WALLET type, though these typescript functions aren't really used in the tests, they are mainly there for debugging/sanity checking. 

The WALLET signature can be anything, but when it's compatible with our `ERC1271Module` module it will mostly contains data like this:
```
[actual signature data][signature type X][signature type WALLET]
```
So it's kind of like a recursion thing. The first time the signature is processed in the meta tx we detect it's a wallet signature. We then strip off the wallet byte and pass the signature to the wallet contract. The wallet contract can do anything with it as it likes, but our `ERC1271Module` module checks the latest byte of the signature again, which can be any signature type (it could even be a WALLET signature again which is passed on to another contract). This way we can support any signature type even if the initial wallet type is WALLET.